plugins/jobtype/src/azkaban/jobtype/HadoopSparkJob.java

```diff
@@ -0,0 +1,422 @@
+      getLog().info("Need to proxy. Getting tokens.");
+      // get tokens in to a file, and put the location in props
+      Props props = new Props();
+      props.putAll(getJobProps());
+      props.putAll(getSysProps());
+      tokenFile = getHadoopTokens(props);
+      getJobProps().put("env." + HADOOP_TOKEN_FILE_LOCATION, tokenFile.getAbsolutePath());
+    }
+
+    try {
+      super.run();
+    } catch (Exception e) {
+      e.printStackTrace();
+      getLog().error("caught exception running the job");
+      throw new Exception(e);
+    } catch (Throwable t) {
+      t.printStackTrace();
+      getLog().error("caught error running the job");
+      throw new Exception(t);
+    } finally {
+      if (tokenFile != null) {
+        cancelHadoopTokens(tokenFile);
+        if (tokenFile.exists()) {
+          tokenFile.delete();
+        }
+      }
+    }
+  }
+
+  private HadoopSecurityManager loadHadoopSecurityManager(Props props, Logger logger)
+          throws RuntimeException {
+
+    Class<?> hadoopSecurityManagerClass = props.getClass(HADOOP_SECURITY_MANAGER_CLASS_PARAM, true,
+            HadoopHiveJob.class.getClassLoader());
+    getLog().info("Loading hadoop security manager " + hadoopSecurityManagerClass.getName());
+    HadoopSecurityManager hadoopSecurityManager = null;
+
+    try {
+      Method getInstanceMethod = hadoopSecurityManagerClass.getMethod("getInstance", Props.class);
+      hadoopSecurityManager = (HadoopSecurityManager) getInstanceMethod.invoke(
+              hadoopSecurityManagerClass, props);
+    } catch (InvocationTargetException e) {
+      getLog().error(
+              "Could not instantiate Hadoop Security Manager "
+                      + hadoopSecurityManagerClass.getName() + e.getCause());
+      throw new RuntimeException(e.getCause());
+    } catch (Exception e) {
+      e.printStackTrace();
+      throw new RuntimeException(e.getCause());
+    }
+
+    return hadoopSecurityManager;
+
+  }
+
+  private void cancelHadoopTokens(File tokenFile) {
+    try {
+      hadoopSecurityManager.cancelTokens(tokenFile, userToProxy, getLog());
+    } catch (HadoopSecurityManagerException e) {
+      e.printStackTrace();
+      getLog().error(e.getCause() + e.getMessage());
+    } catch (Exception e) {
+      e.printStackTrace();
+      getLog().error(e.getCause() + e.getMessage());
+    }
+  }
+
+  protected File getHadoopTokens(Props props) throws HadoopSecurityManagerException {
+
+    File tokenFile = null;
+    try {
+      tokenFile = File.createTempFile("mr-azkaban", ".token");
+    } catch (Exception e) {
+      e.printStackTrace();
+      throw new HadoopSecurityManagerException("Failed to create the token file.", e);
+    }
+
+    hadoopSecurityManager.prefetchToken(tokenFile, props, getLog());
+
+    return tokenFile;
+  }
+
+  @Override
+  protected String getJavaClass() {
+    return HADOOP_SECURE_SPARK_WRAPPER;
+  }
+
+  @Override
+  protected String getJVMArguments() {
```

Comment:
Again, if this is the exact copy, see if we can reuse it by creating a utility method.